server/lib/src/value.rs

```diff
@@ -387,6 +391,8 @@ pub enum PartialValue {
     UiHint(UiHint),
     Passkey(Uuid),
     DeviceKey(Uuid),
+    /// We compare on the value hash
+    Image(String),
```

Comment:
Why the hash? Could there be other ways to compare?

Reply:
Yeah, I was trying to avoid collisions of filename/type/contents, since you and I might upload `logo.png` with different contents for two different things... I didn't know if this was an in-db representation or something else?

Reply:
Well I think it depends how we end up structuring this internally too. Is this going to "inline" the image attribute on the entry itself? For example:

```
oauth2_rs: name
image: <logo>
```

Or is it going to be a generic blob store and it references?

```
oauth2_rs: name
image: <uuid> (or should it be a hash to the image? Would this be a new type of reference?)

image: <hash>
uuid: <uuid>
```

I feel the blob store might be the better approach here ... 🤔 And then yeah, we reference it by hash of the blob contents. Or we just reference it by asset uuid. 

I think the blob store also is better wrt to access controls? That any authenticated user can "fetch" a blob if you know it's hash, but you have to know the hash? Is there any reason we'd want to "hide" logos? I think querying all logos by any user shouldn't be allowed though ....

Reply:
did we ever make a design doc for this?

Reply:
I'd consider: 

* logos of oauth2 RSen to be a secret, the same way we hide OAuth apps to users who don't have access to them, 'cause ... sekrets.
* system logo - which is public
* User profile pics I'm guessing we'll end up with "opt-in public" or maybe a system default with opt out/in public options?

It kinda makes sense to make images separate somehow in the DB/storage, but then that gets into its own complexity journey with "who owns what, and what permissions do we apply to CRUD them?". It'd be handy if they were a foreign reference to save on memory, but inextricably linked so one doesn't end up with unowned blobs in the system storage.

Reply:
Thats true. I think we inline them on entries then. If we ever need to truly save memory, we can de-dupe them another way. I think this inlining also means we can re-use our existing oauth2 permission controls we have where users who are elligble to an rs get read access to limited attrs of it.

Even though this would be authenticated, would it still be cacheable 🤔

Reply:
The size is mostly a `Vec<u8>` so my understanding is that it's shoved off into the heap anyhow, which saves memory things?

We can set the image to be (quickly) cacheable on the client side to save requests for the big object, I'm not sure how the internal cache works for Kanidm :D

Reply:
Yes and no. It's just where the allocation exists. What actually saves memory here is that entries in the kanidm db cache are wrapped with Arc<> and shared by read transactions, so there is "one" allocation of the image in the heap, that many threads can all be sharing and accessing. It's only when the entry is written to that we need to duplicate the content of the image in memory for the new write (copy-on-write afterall). 

So given that oauth2-rs are a "low write" target, then mostly there will just be one copy of the image in memory at a time, associated to any oauth2-rs. Similar for the domain.

The exception here is persons/accounts because they ARE a high frequency write target, because of credential updates and session updates as they login/out. So that may cause some impacts on memory/cloning, but honestly probably nothing too wild. 

I think we do it the simple way like this first, and if it becomes a problem, then we find a way to transparently cache these blobs in a separate manner / storage pool. But that can be added later with migrations etc.